mongomock/collection.py

```diff
@@ -1883,26 +1900,44 @@ def _extend_collection(out_collection, field, expression):
                 elif k == '$limit':
                     out_collection = out_collection[:v]
                 elif k == '$unwind':
-                    if isinstance(v, dict):
-                        raise NotImplementedError(
-                            'Mongomock does not handle parameters as an object yet')
-                    if not isinstance(v, helpers.basestring) or v[0] != '$':
+                    if not isinstance(v, dict):
+                        v = {'path': v}
+                    path = v['path']
+                    if not isinstance(path, helpers.basestring) or path[0] != '$':
                         raise ValueError(
                             "$unwind failed: exception: field path references must be prefixed "
-                            "with a '$' '%s'" % v)
+                            "with a '$' '%s'" % path)
+                    path = path[1:]
+                    should_preserve_null_and_empty = v.get('preserveNullAndEmptyArrays')
+                    include_array_index = v.get('includeArrayIndex')
                     unwound_collection = []
                     for doc in out_collection:
-                        array_value = get_value_by_dot(doc, v[1:])
-                        if array_value in (None, []):
+                        try:
+                            array_value = get_value_by_dot(doc, path)
+                        except KeyError:
+                            if should_preserve_null_and_empty:
+                                unwound_collection.append(doc)
+                            continue
+                        if array_value is None:
+                            if should_preserve_null_and_empty:
+                                unwound_collection.append(doc)
                             continue
-                        elif not isinstance(array_value, list):
-                            raise TypeError(
-                                '$unwind must specify an array field, field: '
-                                '"%s", value found: %s' % (v, array_value))
-                        for field_item in array_value:
-                            unwound_collection.append(copy.deepcopy(doc))
-                            unwound_collection[-1] = set_value_by_dot(
-                                unwound_collection[-1], v[1:], field_item)
+                        if isinstance(array_value, list):
+                            iter_array = enumerate(array_value)
+                        else:
+                            iter_array = [(None, array_value)]
+                        if not array_value and should_preserve_null_and_empty:
```

Comment:
please replace `array_value` by `iter_array`, and add a test to make sure a field with `false` value is preserved.

Reply:
Done. Although iter_array is never falsy.